### PR TITLE
copr: Switch to use nanosecs for Duration

### PR DESCRIPTION
**Pull request by andylokandy**
Signed-off-by: Dian Luo <andylokandy@hotmail.com>

###  What have you changed?

- Removed the fsp in current Duration implementation.
- Changed to use the same memory format for the Duration.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

- Unit test

### CODE REVIEW

**Comment by andylokandy** [comment]
Still ICE...

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 1]
```diff
@@ -60,16 +59,16 @@ fn check_second(second: u32) -> Result<u32> {
...
+        Err(Error::truncated_wrong_val("NANOS", nanos))
     } else {
-        Ok(micros)
+        Ok(nanos)
     }
 }
 
 mod parser {
-    use super::{check_hour, check_minute, check_second, Error, MICRO_WIDTH, TEN_POW};
+    use super::{check_hour, check_minute, check_second, Error, NANO_WIDTH, TEN_POW};
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: Maybe we could upgrade the `nom` to the latest version in the parser.

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 2]
```diff
@@ -244,217 +243,179 @@ mod parser {
...
+            nanos: self.nanos.abs(),
+            ..self
+        }
+    }
 
-        secs + if self.get_neg() { -micros } else { micros } * 1000
+    pub fn from_secs(secs: i64, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
+        Ok(Duration {
+            nanos: secs.checked_mul(NANOS_PER_SEC).ok_or_else(|| {
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: Do we need to check whether it overflows max duration?

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 3]
```diff
@@ -244,217 +243,179 @@ mod parser {
...
+            fsp,
+        })
     }
 
-    /// Constructs a `Duration` from `nanos` with `fsp`
-    pub fn from_nanos(nanos: i64, fsp: i8) -> Result<Duration> {
-        Duration::from_micros(nanos / 1000, fsp)
+    pub fn from_millis(millis: i64, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
+        let nanos = millis
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: ditto

File: components/tidb_query/src/codec/mysql/duration.rs [hunk 4]
```diff
@@ -244,217 +243,179 @@ mod parser {
...
-        let mut hours = (secs / i64::from(SECS_PER_HOUR)) as u32;
-        let mut minutes = (secs % i64::from(SECS_PER_HOUR) / i64::from(SECS_PER_MINUTE)) as u32;
-        let mut secs = (secs % 60) as u32;
-
-        round(&mut hours, &mut minutes, &mut secs, &mut micros, fsp)?;
-        Ok(Duration::new(neg, hours, minutes, secs, micros, fsp))
+        let nanos = micros
+            .checked_mul(NANOS_PER_MICRO)
+            .ok_or_else(|| Error::Eval("DURATION OVERFLOW".to_string(), ERR_DATA_OUT_OF_RANGE))?;
+        let nanos = round(nanos, fsp);
```
Comment (lonng) [inline components/tidb_query/src/codec/mysql/duration.rs]: ditto

File: components/tidb_query/src/codec/mysql/duration.rs (hunk 2 above)
Comment (andylokandy) [reply components/tidb_query/src/codec/mysql/duration.rs]: Done

File: components/tidb_query/src/codec/mysql/duration.rs (hunk 1 above)
Comment (andylokandy) [reply components/tidb_query/src/codec/mysql/duration.rs]: Maybe later. Otherwise we have to also refactor the time module.

**Pull request merged**